File: music-commons-core/src/main/java/net/transgressoft/commons/music/playlist/PlaylistNodeAttribute.java
package net.transgressoft.commons.music.playlist;

import net.transgressoft.commons.music.audio.AudioItem;
import net.transgressoft.commons.query.BooleanQueryTerm;
import net.transgressoft.commons.query.EntityAttribute;

public enum PlaylistNodeAttribute implements EntityAttribute<MutablePlaylist<AudioItem>> {

    SELF;

    public BooleanQueryTerm<AudioPlaylist<AudioItem>> isDirectory() {
        return queryEntity -> queryEntity.getAttribute(this).isDirectory();
    }

    public BooleanQueryTerm<AudioPlaylist<AudioItem>> isNotDirectory() {
        return queryEntity -> ! queryEntity.getAttribute(this).isDirectory();
    }

    public BooleanQueryTerm<AudioPlaylist<AudioItem>> audioItemsAnyMatch(BooleanQueryTerm<AudioItem> queryPredicate) {
        return queryEntity -> queryEntity.getAttribute(SELF).audioItemsAnyMatch(queryPredicate);
    }

    public BooleanQueryTerm<AudioPlaylist<AudioItem>> audioItemsAllMatch(BooleanQueryTerm<AudioItem> queryPredicate) {
        return queryEntity -> queryEntity.getAttribute(SELF).audioItemsAllMatch(queryPredicate);
    }
}

File: music-commons-core/src/main/kotlin/net/transgressoft/commons/music/playlist/PlaylistNodeAttribute.kt
package net.transgressoft.commons.music.playlist

import net.transgressoft.commons.music.audio.AudioItem
import net.transgressoft.commons.query.BooleanQueryTerm
import net.transgressoft.commons.query.EntityAttribute

enum class PlaylistNodeAttribute : EntityAttribute<AudioPlaylist<AudioItem>> {
    SELF;

    val isDirectory: BooleanQueryTerm<AudioPlaylist<AudioItem>>
        get() = BooleanQueryTerm { queryEntity: AudioPlaylist<AudioItem> -> queryEntity.getAttribute(this).isDirectory }
    val isNotDirectory: BooleanQueryTerm<AudioPlaylist<AudioItem>>
        get() = BooleanQueryTerm { queryEntity: AudioPlaylist<AudioItem> -> !queryEntity.getAttribute(this).isDirectory }

    fun audioItemsAnyMatch(queryPredicate: BooleanQueryTerm<AudioItem>): BooleanQueryTerm<AudioPlaylist<AudioItem>> {
        return BooleanQueryTerm { queryEntity: AudioPlaylist<AudioItem> ->
            queryEntity.getAttribute(SELF).audioItemsAnyMatch(queryPredicate)
        }
    }

    fun audioItemsAllMatch(queryPredicate: BooleanQueryTerm<AudioItem>): BooleanQueryTerm<AudioPlaylist<AudioItem>> {
        return BooleanQueryTerm { queryEntity: AudioPlaylist<AudioItem> ->
            queryEntity.getAttribute(SELF).audioItemsAllMatch(queryPredicate)
        }
    }
}